### PR TITLE
fix: avoid dune-local opam lock convoys

### PR DESCRIPTION
**Pull request by jeong-sik**
## Summary
- Acquire the local Dune throttle before taking the shared opam switch lock so queued builds do not block opam pin repair while waiting.
- Add bounded post-Dune opam-lock wait via `MASC_OPAM_LOCK_AFTER_DUNE_TIMEOUT`, releasing the Dune lock if older lock-order invocations are still holding opam.
- Suppress duplicate missing-lock-tool warnings on the opam-lock path.
- Extend dune-local script tests and source hardening guards for the lock ordering and edge cases.

## Verification
- `bash -n scripts/dune-local.sh`
- `git diff --check origin/main...HEAD`
- `scripts/check-oas-pin.sh --local-only`
- `env MASC_OPAM_LOCK=0 scripts/dune-local.sh build test/test_dune_local_script.exe test/test_ci_hardening_source.exe`
- `./_build/default/test/test_dune_local_script.exe` (24/24)
- `./_build/default/test/test_ci_hardening_source.exe` (53/53)

Refs #13827: current-main stale fleet duplicate was repaired separately in #13830 before this rebase.

### CODE REVIEW

File: scripts/dune-local.sh [hunk 1]
```diff
@@ -125,13 +152,39 @@ _needs_opam_lock() {
...
 if _needs_opam_lock; then
-  script_path="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)/$(basename "${BASH_SOURCE[0]}")"
   printf '[dune-local] waiting for opam switch lock %s\n' "$opam_lock_path" >&2
   if command -v lockf >/dev/null 2>&1; then
+    if [[ "${MASC_DUNE_LOCK_HELD:-0}" = "1" \
+          && "${MASC_OPAM_LOCK_AFTER_DUNE_TIMEOUT:-60}" != "0" ]]; then
+      opam_lock_timeout="${MASC_OPAM_LOCK_AFTER_DUNE_TIMEOUT:-60}"
+      if ! [[ "$opam_lock_timeout" =~ ^[0-9]+$ ]]; then
+        printf '[dune-local] invalid MASC_OPAM_LOCK_AFTER_DUNE_TIMEOUT=%q; expected non-negative integer seconds\n' \
+          "$opam_lock_timeout" >&2
```
[the remark this answers is not in the document]
Comment (jeong-sik) [reply scripts/dune-local.sh]: Addressed in 52918db635 and covered in 1c89e018dd: MASC_OPAM_LOCK_AFTER_DUNE_TIMEOUT is now opt-in only; unset or 0 preserves the historical wait-forever behavior. Added a regression that verifies lockf is not called with -t when the env is unset. Verified test_dune_local_script.exe and git diff --check.

File: scripts/dune-local.sh [hunk 2]
```diff
@@ -125,13 +152,39 @@ _needs_opam_lock() {
...
+      if [[ "$status" -eq 75 ]]; then
+        printf '[dune-local] opam switch lock stayed busy for %ss after acquiring Dune lock; releasing Dune lock to avoid mixed lock-order deadlock\n' \
+          "$opam_lock_timeout" >&2
+        printf '[dune-local] retry after older dune-local invocations drain, or set MASC_OPAM_LOCK_AFTER_DUNE_TIMEOUT=0 to wait indefinitely\n' >&2
+      fi
+      exit "$status"
+    fi
     exec lockf -k "$opam_lock_path" env MASC_OPAM_LOCK_HELD=1 "$script_path" "$@"
   elif command -v flock >/dev/null 2>&1; then
     exec flock "$opam_lock_path" env MASC_OPAM_LOCK_HELD=1 "$script_path" "$@"
```
[the remark this answers is not in the document]
Comment (jeong-sik) [reply scripts/dune-local.sh:218]: Addressed in 52918db635 and covered in 1c89e018dd: the flock path now uses flock -w when the bounded opam-lock timeout is explicitly enabled, with the same Dune-lock release diagnostic as lockf. Added a flock timeout regression. Verified test_dune_local_script.exe and git diff --check.

File: test/test_dune_local_script.ml [hunk 3]
```diff
@@ -308,13 +333,171 @@ exec "$@"
...
+          ~env:
+            [
+              ("PATH", bin_dir);
+              ("GIT_CEILING_DIRECTORIES", dir);
+              ("DUNE_LOCAL_LOCK", Filename.concat dir "dune-local.lock");
+              ("MASC_SKIP_PIN_CHECK", "1");
+              ("MASC_SKIP_DEPS_CHECK", "1");
+              ("MASC_SKIP_OCAML_VERSION_CHECK", "1");
+            ]
+          ~unset_env:[ "GITHUB_ACTIONS"; "MASC_DUNE_LOCK_HELD" ]
```
[the remark this answers is not in the document]
Comment (jeong-sik) [reply test/test_dune_local_script.ml:562]: Addressed in 1c89e018dd: the missing-lock-tools test now keeps the fake opam on PATH and asserts the opam-lock warning is suppressed, so the test covers the intended branch rather than only the no-opam path. Verified test_dune_local_script.exe and git diff --check.

File: scripts/dune-local.sh [hunk 4]
```diff
@@ -125,13 +152,39 @@ _needs_opam_lock() {
...
-  else
+  elif [[ "${MASC_DUNE_LOCK_HELD:-0}" != "1" ]]; then
+    # Skip the warning when the Dune-lock branch above already emitted
+    # an equivalent "neither lockf nor flock found" message in this
+    # process. Without the guard, both paths print and operators see
+    # two warnings even though there is only one underlying problem.
     printf '[dune-local] warning: neither lockf nor flock found; opam switch checks are unlocked\n' >&2
   fi
```
[the remark this answers is not in the document]
Comment (jeong-sik) [reply scripts/dune-local.sh]: Addressed in 52918db635 and covered in 1c89e018dd: the opam-lock missing-tool warning is now suppressed via the local dune_lock_warning_emitted flag, not MASC_DUNE_LOCK_HELD. The warning-once test now explicitly asserts the opam warning is absent. Verified test_dune_local_script.exe and git diff --check.

File: test/test_ci_hardening_source.ml [hunk 5]
```diff
@@ -560,6 +560,13 @@ let test_oas_pin_source_contracts () =
   check bool "dune-local.sh exposes shared opam switch lock path" true
     (file_contains_pattern "scripts/dune-local.sh"
        "MASC_OPAM_LOCK_PATH:-/tmp/me-opam-switch.lock");
+  check bool "dune-local.sh takes build lock before opam switch lock" true
+    (match
+       file_pattern_position "scripts/dune-local.sh" "_needs_dune_lock",
+       file_pattern_position "scripts/dune-local.sh" "_needs_opam_lock"
```
[the remark this answers is not in the document]
Comment (jeong-sik) [reply test/test_ci_hardening_source.ml]: Addressed in 975251e690: the source hardening assertion now anchors on the actual lock-acquisition log messages (waiting for lock before waiting for opam switch lock), not the helper function definition order. Verified test_ci_hardening_source.exe, test_dune_local_script.exe, and git diff --check.

File: scripts/dune-local.sh [hunk 6]
```diff
@@ -125,13 +152,76 @@ _needs_opam_lock() {
...
+  if [[ "${MASC_DUNE_LOCK_HELD:-0}" = "1" \
+        && -n "${MASC_OPAM_LOCK_AFTER_DUNE_TIMEOUT:-}" \
+        && "${MASC_OPAM_LOCK_AFTER_DUNE_TIMEOUT}" != "0" ]]; then
+    opam_lock_timeout="${MASC_OPAM_LOCK_AFTER_DUNE_TIMEOUT}"
+    if ! [[ "$opam_lock_timeout" =~ ^[0-9]+$ ]]; then
+      printf '[dune-local] invalid MASC_OPAM_LOCK_AFTER_DUNE_TIMEOUT=%q; expected non-negative integer seconds\n' \
+        "$opam_lock_timeout" >&2
+      exit 2
+    fi
+    opam_bounded_wait=1
```
[the remark this answers is not in the document]
Comment (jeong-sik) [reply scripts/dune-local.sh]: Addressed in 975251e690: the timeout env is now validated as numeric first, normalized as base-10, and only enables bounded wait when the normalized value is > 0. Added a regression for MASC_OPAM_LOCK_AFTER_DUNE_TIMEOUT=00 preserving wait-forever behavior. Verified test_dune_local_script.exe, test_ci_hardening_source.exe, and git diff --check.

File: test/test_ci_hardening_source.ml (hunk 5 above)
[the remark this answers is not in the document]
Comment (jeong-sik) [reply test/test_ci_hardening_source.ml]: Addressed in current head 975251e690: the hardening assertion now anchors on the actual lock-wait messages, including the Dune lock wait preceding the opam switch lock wait, rather than comparing helper definition names. Verified bash -n scripts/dune-local.sh, ocamlc parse for test/test_ci_hardening_source.ml, and git diff --check origin/main...HEAD.
